transport/http/do_test.go

```diff
@@ -0,0 +1,355 @@
+		timeout time.Duration
+	}{
+		{
+			desc:    "GET error",
+			method:  "GET",
+			path:    "/",
+			timeout: 0,
+		},
+		{
+			desc:    "POST error",
+			method:  "POST",
+			path:    "/",
+			timeout: 0,
+		},
+	}
+
+	httpTransport := NewTransport()
+
+	for _, tt := range tests {
+		server := httptest.NewServer(http.HandlerFunc(
+			func(w http.ResponseWriter, r *http.Request) {
+				http.Error(w, "Error out", http.StatusInternalServerError)
+			},
+		))
+		defer server.Close()
+
+		out := httpTransport.NewSingleOutbound(server.URL)
+
+		require.NoError(t, out.Start(), "failed to start outbound")
+		defer out.Stop()
+
+		ctx := context.Background()
+		ctx, cancel := context.WithTimeout(ctx, tt.timeout)
+		defer cancel()
+
+		req, err := http.NewRequest(tt.method, server.URL, bytes.NewReader([]byte("world")))
+		req = req.WithContext(ctx)
+
+		client := http.Client{Transport: out}
+
+		res, err := client.Do(req)
+
+		assert.Error(t, err)
+		assert.Nil(t, res)
+	}
+}
+
+func TestHTTPApplicationError(t *testing.T) {
+	tests := []struct {
+		desc     string
+		status   string
+		appError bool
+	}{
+		{
+			desc:     "ok",
+			status:   "success",
+			appError: false,
+		},
+		{
+			desc:     "error",
+			status:   "error",
+			appError: true,
+		},
+		{
+			desc:     "not an error",
+			status:   "lolwut",
+			appError: false,
+		},
+	}
+
+	httpTransport := NewTransport()
+
+	for _, tt := range tests {
+		server := httptest.NewServer(http.HandlerFunc(
+			func(w http.ResponseWriter, r *http.Request) {
+				w.Header().Add("Status", tt.status)
+				defer r.Body.Close()
+			},
+		))
+		defer server.Close()
+
+		out := httpTransport.NewSingleOutbound(server.URL)
+
+		require.NoError(t, out.Start(), "failed to start outbound")
+		defer out.Stop()
+
+		ctx := context.Background()
+		ctx, cancel := context.WithTimeout(ctx, 100*testtime.Millisecond)
+		defer cancel()
+
+		req, err := http.NewRequest("GET", server.URL, bytes.NewReader([]byte("world")))
+		req = req.WithContext(ctx)
+
+		client := http.Client{Transport: out}
+
+		res, err := client.Do(req)
+
+		assert.Equal(t, res.Header.Get("Status"), tt.status, "%v: application status", tt.desc)
+
+		if !assert.NoError(t, err, "%v: call failed", tt.desc) {
+			continue
+		}
+
+		if !assert.NoError(t, res.Body.Close(), "%v: failed to close response body") {
+			continue
+		}
+	}
+}
+
+func TestHTTPCallFailures(t *testing.T) {
+	notFoundServer := httptest.NewServer(http.NotFoundHandler())
+	defer notFoundServer.Close()
+
+	internalErrorServer := httptest.NewServer(http.HandlerFunc(
+		func(w http.ResponseWriter, r *http.Request) {
+			http.Error(w, "great sadness", http.StatusInternalServerError)
+		}))
+	defer internalErrorServer.Close()
+
+	httpTransport := NewTransport()
+
+	tests := []struct {
+		url      string
+		messages []string
+	}{
+		{"not a URL", []string{"protocol scheme"}},
+		{notFoundServer.URL, []string{"404", "page not found"}},
+		{internalErrorServer.URL, []string{"great sadness"}},
```

Comment:
Can we add a test for a fully qualified URL like "http://example.com/". This should fail because we only support relative paths with this RoundTripper implementation.